### ISSUE TITLE
Some clarification on Barracuda tensors rank fixed at 4

### ISSUE THREAD

**Issue by zaino** [Unity-Technologies/barracuda-release#166]
Hello!
I have a model that works fine in PyTorch, but when I run it in Unity, it’s not behaving like it’s supposed to.

I think the problem is this: one of the input tensor is a mask of 0 or 1 (only one is ever 1), kept as floats to make later operations easier for me. Now, in PyTorch that tensor is supposed to be of rank 2, shape(1,4). I then transform it into an index with torch.nonzero( mask[0] ).

I suspect that since in Barracuda tensor rank is always 4, this is returning 0 as index. I thought of squeezing the mask to rank 2, but as per documentation, Barracuda doesn’t support squeeze with dim<=1.

To be honest, the actual solution to this it’s not that important to me, I already have alternatives that work, but I would like to understand if my hunch is correct, because I have a feeling I will have a lot of issues porting my models in Barracuda due to Barracuda having tensor rank fixed at 4.

If any of the devs could suggest an elegant way of solving this issue though, I think that will give me the insight I need to work reasonably around the rank 4 limitation.

Thanks in advance for the assistance!

**Comment by AlexRibard** [comment]
Hi! @zaino can you share the model you are having issues with?
You are correct, Barracuda operates on rank4 (NHWC) tensors.
```
rank1 : N,1,1,1
rank2 : N,1,1,C
rank3 : N,1W,C
rank4 : N,H,W,C
```
things gets a bit harry coming from Pytorch as we need to transpose NCHW (native pytroch) to NHWC Barracuda.
The final mapping is as follows
```
Pytroch                  Barracuda
rank1 : N       -> N,1,1,1
rank2 : N,C     -> N,1,1,C
rank3 : N,C,W   -> N,1,W,C
rank4 : N,C,H,W -> N,H,W,C
```
I hope that makes sense?

**Comment by fabiozinno** [comment]
Hi Alex, thanks for the quick reply!
Unfortunately I cannot share the onnx model as is, since it's work stuff. 
I'll see if I can whip up a similar toy example to share.

What you said makes sense and gives me some more insight.
To give you more detail, the reason why I need an index is because my model has 4 sub-modules in a ModuleList, and based on the mask, I want to call inference on the right one.
Exporting a model that forces the index in the module list to be a constant (I tried 0,1,2,3 and they all behave as I expect) works in Unity, but whenever I try to get the index from the float mask (e.g. [0, 1, 0, 0]), it always selects module[0].

Trying to figure out why this happens,  I'm listing the possible points of failure in my head:
1) **the different rank** handling could be a problem, and your info helps to understand this. BTW looking at the documentation for Barracuda, it seems like the limitation of squeeze/unsqueeze dimension <=1 is not listed anymore for 1.3.3-preview. Is that an omission or is it actually supported now?
2) **torch.nonzero** might not work in Barracuda/ONNX. I see in the documentation: 
_"Notes: Known bug for non const inputs with at least one dimension of size 1."_ 
If you know, could you elaborate? I am using nonzero to convert the input mask into an index.
3) I'm also not sure if using the content of a tensor as an index is a valid operation, even if it's int type like the return value of torch.nonzero. It works in pytorch, not sure if it works in Barracuda.

Last question: is there a way to have the model output print messages that would show in Unity's log window?

Thanks again for the help!

**Comment by zaino** [comment]
Hi Alex, thanks for the quick reply!
Unfortunately I cannot share the ONNX model as is, since it's work stuff.
This is the snippet code for the model's forward():

```
def forward(self, data, masks):
    selmask = masks[..., 0:4]
    index = torch.nonzero( selmask )[..., -1]
    model_ = self.models[index]
    ....
```

This works in pytorch, and the right model is selected. In pytorch, the shape of _masks_ is (1, 8).
In Unity, it appears _index_ ends up always being 0. 
To test if the problem lies with the shape of masks in Barracuda or in torch.nonzero not working as expected, I tried hardcoding _selmask_ like this:

```
def forward(self, data, masks):
    selmask = torch.tensor( [[0,0,1,0]] ).to(self.device)
    index = torch.nonzero( selmask )[..., -1]
    model_ = self.models[index]
    ....
```

and in this case, _index=2_ in Unity as well, like it should.

For completeness, this is the C# code unity side:
```
var inputs = new Dictionary<string, Tensor>();
int[] dataShape = new int[] {1, 1, 1, dataSize };
inputs["data"] = new Tensor(dataShape , data, "data");
int[] masksShape = new int[] {1, 1, 1, 8 };
float[] maskData = new float[8]  {  0.0f, 0.0f, 1.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f };
inputs["masks"] = new Tensor(masksShape, maskData, "masks");
Worker.Execute(inputs); 
```

If there was a way to print out the tensor shape and content as they are when they get in Barracuda, and see that in the Unity Log, the problem would probably be evident right away.

Hopefully this gives you more info to suggest a possible solution.

Thanks!

**Comment by zaino** [comment]
Quick update, there obviously is a way to peek at the tensors in Barracuda, and it's to make the model output them :)
So I did it:

```
def forward(self, data, masks):
    selmask = masks[..., 0:4]
    index = torch.nonzero( selmask )[..., -1]
    model_ = self.models[index]
    ....
    return out, index, selmask
```

and when peeking at them in Unity, for a `selmask = [0, 0, 1, 0]` coming from Unity,  ` index = 2`, like it should.
Unfortunately though, the model is not behaving like it does when I hardcode `selmask = torch.tensor( [[0,0,1,0]] )` on the python side. It really behaves like somehow `self.models[0]` is being selected.

At this point, I'm very confused.

**Comment by AlexRibard** [comment]
Additionally to query intermediate layers without specifically changing your model, you can call 
```
    /// <param name="additionalOutputs">the additional outputs to track but not directly specified by the model</param>
    /// <param name="trimOutputs">by specifying this list of outputs, all other non-specified outputs will be discarded</param>
CreateWorker(Type type, Model model, string[] additionalOutputs, string[] trimOutputs)
```

**Comment by AlexRibard** [comment]
@zaino given that your model is private IP, I think it will be easier if we take this offline.
Our support email is barracuda-support@unity3d.com, feel free to email us.
